Initial implementation of vector similarity index

rschu1ze opened the pull request:
This PR changed too many things to enumerate here. I tried to keep individual commits small, self-contained, and properly described (commit msg). Reviewers may go through the individual commits.

In sum:
- the old (experimental) vector indexes `annoy` and `usearch` are gone (*)
- they were replaced by a new (still experimental) index type `vector_similarity`, 
- obscure edge cases of vector search are no longer supported: tuples, WHERE-type queries, Lp norms
- polishing and fixes all over the place
- the new `vector_similarity` index is under setting `allow_experimental_vector_similarity_index`

(*) it will still be possible to load old tables `annoy`/`usearch` indexes but data insert / search will throw a message that recommends migration to `vector_similarity`.

Fixes: #49669

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

rschu1ze committed:
Un-screw usearch's build description

No directory 'SimSIMD-map' exists, the build only worked because SimSIMD
support in usearch was (accidentally?) disabled. This commit corrects
the build description. SimSIMD support in usearch will be enabled by a
later commit.

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #63675 from rschu1ze/ClickHouse:vector-search -->
*This is an automated comment for commit 6b9b616960ab965e72e88949fdfad7341df17936 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63675/6b9b616960ab965e72e88949fdfad7341df17936/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

aaroncueckermann commented:
@rschu1ze Sorry to jump into the draft, I was just wondering if you had an idea about what kind of search speed performance improvements we might see in Clickhouse Cloud (if any) due to the SimSIMD hardware acceleration (e.g. negligible, 2x, 10x?)

rschu1ze commented:
Discussion in community Slack: https://clickhousedb.slack.com/archives/CU170QE9H/p1717514940423209?thread_ts=1717408884.549689&cid=CU170QE9H

rschu1ze committed:
Add migration helper for legacy 'annoy' and 'usearch' indexes types

Index types 'annoy' and 'usearch' were removed and replaced by
'vector_similarity' indexes in an earlier commit.

This means unfortuantely, that if customers have tables with these
indexes and upgrade, their database might not start anymore - the
system loads the metadata at startup, thinks something is wrong with
such tables, and halts immediately.

This commit adds support for loading and attaching such indexes back.
Data insert or use (search) return an error which recommends a migration
to 'vector_similarity' indexes. The implementation is generally similar
to what has recently been implemented for 'full_text' indexes [1, 2].

[1] https://github.com/ClickHouse/ClickHouse/pull/64656
[2] https://github.com/ClickHouse/ClickHouse/pull/64846

antaljanosbenjamin reviewed (commented):
Review until [53a6633](https://github.com/ClickHouse/ClickHouse/pull/63675/commits/53a6633849278022ec25ef066d306e1b14039e27)

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexUSearch.cpp:
```diff
@@ -433,7 +401,7 @@ void usearchIndexValidator(const IndexDescription & index, bool /* attach */)
         throw Exception(
             ErrorCodes::ILLEGAL_COLUMN,
-            "USearch can only be created on columns of type Array(Float32) and Tuple(Float32[, Float32[, ...]])");
+            "USearch indexes can only be created on columns of type Array(Float32))");
```
Done, tnx.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.cpp:
```diff
@@ -0,0 +1,513 @@
+    }
+    else if constexpr (is_unordered_map<T>)
+    {
+        String joined_keys;
+        for (const auto & [k, _] : t)
+        {
+            if (!joined_keys.empty())
+                joined_keys += ", ";
+            joined_keys += k;
+        }
+        return joined_keys;
```
It works but not in ClickHouse. Our libcxx is too old :-( I'll add at least a TODO comment.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.cpp:
```diff
@@ -265,11 +268,11 @@ std::vector<size_t> MergeTreeIndexConditionVectorSimilarity::getUsefulRangesImpl
     const std::vector<float> reference_vector = condition.getReferenceVector();
 
-    const auto granule = std::dynamic_pointer_cast<MergeTreeIndexGranularityVectorSimilarity<Metric>>(idx_granule);
+    const auto granule = std::dynamic_pointer_cast<MergeTreeIndexGranuleVectorSimilarity>(idx_granule);
```
That's a valid point.

The check that the index distance function == the query distance function is in file "src/Storages/MergeTree/VectorSimilarityCondition.cpp", function `VectorSimilarityCondition::alwaysUnknownOrTrue`. Just checked what happens if there is a mismatch, and (as expected) the index is simply not used. In that sense, I think everything is just fine.

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/annindexes.md:
````diff
@@ -27,75 +27,85 @@ Function `Distance` computes the distance between two vectors. Often, the Euclid
 ORDER BY id;
 ```
 
+Parameters:
````
Did not think about that but I implemented it like you said. Also added a concrete index creation example to the docs (I plan to rework the docs anyways).

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.cpp:
```diff
@@ -0,0 +1,516 @@
+{
+
+/// The only index type currently supported by USearch
+std::set<String> methods = {"hnsw"};
+
+/// Maps from user-facing name to internal name
+std::unordered_map<String, unum::usearch::metric_kind_t> distanceFunctionToMetricKind = {
+    {VectorSimilarityCondition::L2Distance, unum::usearch::metric_kind_t::l2sq_k},
+    {VectorSimilarityCondition::CosineDistance, unum::usearch::metric_kind_t::cos_k},
+    {VectorSimilarityCondition::DotProduct, unum::usearch::metric_kind_t::ip_k}};
+
+/// Maps from user-facing name to internal name
+/// Most of them are commented out because usearch goes boom, see tests/queries/0_stateless/02354_vector_search_hnsw_quantization.sql
+std::unordered_map<String, unum::usearch::scalar_kind_t> quantizationToScalarKind = {
```
Constness for the win! Done

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.h:
```diff
@@ -88,6 +110,8 @@ struct MergeTreeIndexGranuleVectorSimilarity final : public IMergeTreeIndexGranu
     /// outside usearch exists which we should keep it separately versioned.
     static constexpr UInt64 FILE_FORMAT_VERSION = 1;
+
+    LoggerPtr logger = getLogger("VectorSimilarityIndex");
```
The idea is charming. On the other hand, the overall convention seems to keep logger names generic and free of things like the query id, table names, etc.

To associate the log with queries or indexes, one could use the thread id which is afaik part of the log.

rschu1ze commented on tests/queries/0_stateless/02354_vector_search_multiple_indexes.sql:
```diff
@@ -0,0 +1,14 @@
+ALTER TABLE tab ADD INDEX vec_idx1(vec) TYPE vector_similarity('hnsw', 'cosineDistance');
+ALTER TABLE tab ADD INDEX vec_idx2(vec) TYPE vector_similarity('hnsw', 'L2Distance'); -- silly but creating the same index also works for non-vector indexes ...
+
+DROP TABLE tab;
```
Sure

rschu1ze commented on src/Storages/MergeTree/VectorSimilarityCondition.h:
```diff
@@ -33,8 +33,14 @@ class VectorSimilarityCondition
 {
 public:
     static constexpr auto L2Distance = "L2Distance";
+    static constexpr auto DistanceL2 = "DistanceL2";
```
This code will no longer be needed with the analyzer ... we better don't mess with it without good reason.

rschu1ze committed:
Fix CMake integration of usearch and annoy

Registers usearch and annoy properly via configure_config.cmake and
config.h.in like all other 3rd party libs, instead of (mis)using
target_compile_definitions.

rschu1ze committed:
Fix test results (no analyzer support yet ...)

rschu1ze committed:
Remove Annoy indexes

Annoy indexes fell out of favor in the community, at least when it comes
to vector databases. Such indexes work okay-ish low dimensions but they
suffers badly from a curse of dimensionality which makes them inapt for
a high number of dimensions.

Now that Annoy is gone, issue (*) also disappears and we can drop
'no-ubsan', 'no-cpu-aarch64', and 'no-asan' from tests.

(*) spotify/annoy#456

rschu1ze committed:
Remove tuple support

Indexes for approximate nearest neighbourhood (ANN) search (USearch) can
be build on columns of type Array(Float32) or Tuple(Float32[, Float32[, ...]]).
In practice, Arrays(Float32) is the only relevant data type.
Arrays store high-dimensional embeddings consecutively (--> cache
locality) and the additional flexibility of different data types in a
tuple is not needed for vector search.

Therefore removing support for ANN indexes over tuple columns to
simplify the code, tests and docs.

rschu1ze committed:
Remove support code for Lp norm in vector search

It is a generalization of other norms, too expensive to calculate and
not relevant in practice. Also, Usearch doesn't support it.

rschu1ze committed:
Remove support for WHERE-type queries

These kind of vector search similarity queries are rather obscure and
rare in practice. They require the user to specify a maximum distance
which is not intuitive to obtain. Furthermore, these queries are not
natively supported in USearch, so the vector search index had to emulate
these queries.

Therefore simplifying the code base and restricting vector search to
ORDER-BY queries only.

rschu1ze committed:
Add detach/attach test

rschu1ze committed:
Better handling of errors during add, search, and save

rschu1ze committed:
Add a similar sanity check as in other skipping indexes

rschu1ze committed:
Cosmetics

rschu1ze committed:
Remove useless templatization

Makes the code cleaner, compile faster, and the binary smaller.

rschu1ze committed:
Addd logging of basic statistics

rschu1ze committed:
Cosmetics

rschu1ze committed:
Introduce version for for index files in persistence

rschu1ze committed:
Cosmetics: variable naming

rschu1ze committed:
Cosmetics: whitespaces

rschu1ze committed:
Cosmetics: use native types/functions

rschu1ze committed:
Cosmetics: function naming

rschu1ze committed:
Cosmetics: whitespaces

rschu1ze committed:
Remove funny typedef

rschu1ze committed:
Cosmetics: metric --> distance_function (for consistent terminology)

rschu1ze committed:
Cosmetics: variable naming

rschu1ze committed:
Cosmetics: ApproximateNearestNeighborIndexesCommon --> VectorSimilarityCondition

rschu1ze committed:
Cosmetics: ApproximateNearestNeighborInformation --> Info + nest in class

rschu1ze committed:
Cosmetics: simplify inheritance hierarchy

rschu1ze committed:
Cosmetics: minor stuff

rschu1ze committed:
Rename index type "usearch" to "vector_similarity"

First, index type "vector_similarity" is more speaking and user-friendly
than "usearch". Second, we should not expose the name of the library
doing the job (usearch). Of course, the docs will continue to mention
usearch (credit where credit is due).

Existing setting `allow_experimental_usearch_index` was marked obsolete.
A new settings `allow_experimental_vector_similarity_index` was added.

rschu1ze committed:
Rework vector index parameters

USearch (similar to FAISS) allows to specify the distance function,
quantization, and various HNSW meta-parameters for index creation and
sarch. Some users wished for greater configurability, so let's expose
them.

Index creation now requires either
- 2 parameters (with the other 4 parameters taking on default values), or
- 6 parameters for full control

This commit also remove quantization `f64` (that would be upsampling).

rschu1ze committed:
Allow un-quoted skip index parameters

Previously, only this syntax to create a skip index worked:

   INDEX index_name column_name TYPE vector_similarity('hnsw', 'L2Distance')

Now, this syntax will work as well:

  INDEX index_name column_name TYPE vector_similarity(hnsw, L2Distance)

rschu1ze commented:
Resolving new conflicts ...

rschu1ze commented:
ClickHouse Stateful tests (tsan, ParallelReplicas): #66683

antaljanosbenjamin commented on src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.cpp:
```diff
@@ -0,0 +1,492 @@
+
+    auto result = index->search(reference_vector.data(), limit);
+    if (result.error)
+        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not search in vector similarity index, error: " + String(result.error.release()));
```
```suggestion
        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not search in vector similarity index, error: {}", String(result.error.release()));
```

antaljanosbenjamin commented on src/Storages/MergeTree/MergeTreeIndexUSearch.cpp:
```diff
@@ -41,22 +41,37 @@ namespace ErrorCodes
 namespace
 {
 
+std::unordered_map<String, unum::usearch::metric_kind_t> nameToMetricKind = {
```
`const` for president! I think here it actually makes sense, because we don't want to change this map.

antaljanosbenjamin commented on src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.cpp:
```diff
@@ -0,0 +1,492 @@
+    auto result = Base::load_from_stream(callback);
+    if (result.error)
+        /// See the comment in MergeTreeIndexGranuleVectorSimilarity::deserializeBinary why we throw here
+        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not load vector similarity index, error: " + String(result.error.release()) + " Please drop the index and create it again.");
```
```suggestion
        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not load vector similarity index, error: {} Please drop the index and create it again."", String(result.error.release()));
```

For the other exceptions too.

antaljanosbenjamin reviewed (approved):
Only minor things, feel free to merge the PR and the issues can be addressed in another PR easily, none of them critical.

rschu1ze commented:
Thanks @antaljanosbenjamin Thanks for checking. I actually already work on the [successor PR](https://github.com/ClickHouse/ClickHouse/pull/68248/commits) and will include all your suggestions there.
